### PR TITLE
[ADD] new module sale_invoice_blocking

### PR DESCRIPTION
**Pull request by damdam-s**
allow to block the invoicing of a sale order

### CODE REVIEW

File: sale_invoice_blocking/models/invoice_blocking_reason.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import fields, models
+
+
+class InvoiceBlockingReason(models.Model):
+    _name = "invoice.blocking.reason"
+    _description = "Sale invoice blocking reason"
+
+    name = fields.Char(string="Motif")
```
Comment (rousseldenis) [inline sale_invoice_blocking/models/invoice_blocking_reason.py]:
```suggestion
    name = fields.Char(string="Reason")
```

File: sale_invoice_blocking/views/sale_order.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        <field name="inherit_id" ref="sale.view_order_form" />
+        <field name="arch" type="xml">
+            <field name="client_order_ref" position="after">
+                <field name="invoice_blocking_reason_id" />
+            </field>
+            <xpath
+                expr="//button[@name='%(sale.action_view_sale_advance_payment_inv)d'][contains(concat(' ',normalize-space(@class),' '),' btn-primary ')]"
+                position="attributes"
+            >
+                <attribute
```
Comment (rousseldenis) [inline sale_invoice_blocking/views/sale_order.xml]: @damdam-s Why not modifying invoice_status value computation to 'no' if invoice_blocking_reason_id is set that avoid doing some attrs modifications in view ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review. At least string modification

File: sale_invoice_blocking/i18n/fr.po [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+#. module: sale_invoice_blocking
+#: model:ir.model.fields,field_description:sale_invoice_blocking.field_invoice_blocking_reason__write_date
+msgid "Last Updated on"
+msgstr "Dernière modification le"
+
+#. module: sale_invoice_blocking
+#: model:ir.model.fields,field_description:sale_invoice_blocking.field_invoice_blocking_reason__name
+msgid "Motif"
+msgstr ""
```
Comment (jcoux) [inline sale_invoice_blocking/i18n/fr.po]:
```suggestion
msgid "Reason"
msgstr "Motif"
```

File: sale_invoice_blocking/i18n/fr.po [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+#. module: sale_invoice_blocking
+#: model:ir.model,name:sale_invoice_blocking.model_invoice_blocking_reason
+msgid "Sale invoice blocking reason"
+msgstr "Motifs de blocage de la facturation"
```
Comment (jcoux) [inline sale_invoice_blocking/i18n/fr.po]:
```suggestion
msgstr "Motif de blocage de la facturation"
```

File: sale_invoice_blocking/models/invoice_blocking_reason.py (hunk 1 above)
Comment (jcoux) [inline sale_invoice_blocking/models/invoice_blocking_reason.py]:
```suggestion
    name = fields.Char(string="Motif", required=True)
```
And maybe add a unique SQL constrain?

File: sale_invoice_blocking/readme/USAGE.rst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+To use this module, you need to:
+
+#. Create a new sales order and provide a 'Invoicing block reasons'.
```
Comment (jcoux) [inline sale_invoice_blocking/readme/USAGE.rst]:
```suggestion
#. Create a new sale order and provide a 'Blocking for invoicing'.
```

File: sale_invoice_blocking/views/sale_invoice_blocking_reason_view.xml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+                <field name="name" />
+            </tree>
+        </field>
+    </record>
+    <record id="invoice_blocking_reason_action" model="ir.actions.act_window">
+        <field name="name">Invoicing block reasons</field>
+        <field name="type">ir.actions.act_window</field>
+        <field name="res_model">invoice.blocking.reason</field>
+        <field name="search_view_id" ref="invoice_blocking_reason_search_view" />
+        <field name="view_id" ref="invoice_blocking_reason_list_view" />
```
Comment (jcoux) [inline sale_invoice_blocking/views/sale_invoice_blocking_reason_view.xml:40]:
```suggestion
        <field name="view_id" ref="invoice_blocking_reason_list_view" />
        <field name="view_mode">tree,form</field>
```

**Review by jcoux — CHANGES_REQUESTED** [review]

File: sale_invoice_blocking/views/sale_order.xml (hunk 2 above)
Comment (tarbib) [reply sale_invoice_blocking/views/sale_order.xml]: Thank you @rousseldenis for the suggestion. It's a very good idea indeed.

File: sale_invoice_blocking/models/invoice_blocking_reason.py [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+from odoo import fields, models
+
+
+class InvoiceBlockingReason(models.Model):
+    _name = "invoice.blocking.reason"
+    _description = "Sale invoice blocking reason"
+
+    name = fields.Char(string="Reason", required=True)
```
Comment (jcoux) [inline sale_invoice_blocking/models/invoice_blocking_reason.py:10]: Maybe add SQL unique constrain?

File: sale_invoice_blocking/views/sale_order.xml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+                >{'invisible': ['|', ('invoice_status', '!=', 'to invoice'), ('invoice_blocking_reason_id', '!=', False)]}</attribute>
+            </xpath>
+            <xpath
+                expr="//button[@name='%(sale.action_view_sale_advance_payment_inv)d'][not(contains(concat(' ',normalize-space(@class),' '),' btn-primary '))]"
+                position="attributes"
+            >
+                <attribute
+                    name="attrs"
+                >{'invisible': ['|', '|', ('invoice_status', '!=', 'no'), ('state', '!=', 'sale'), ('invoice_blocking_reason_id', '!=', False)]}</attribute>
+            </xpath-->
```
Comment (jcoux) [inline sale_invoice_blocking/views/sale_order.xml]:
```suggestion
```

File: sale_invoice_blocking/models/sale_order.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        unconfirmed_orders = self.filtered(
+            lambda so: so.state not in ['sale', 'done'])
+        unconfirmed_orders.invoice_status = 'no'
+        confirmed_orders = self - unconfirmed_orders
+        if not confirmed_orders:
+            return
+
+        for order in confirmed_orders:
+            if order.invoice_blocking_reason_id:
+                order.invoice_status = 'no'
```
Comment (jcoux) [inline sale_invoice_blocking/models/sale_order.py]: Suggest more easy (as we have in sale_coupon or sale-workflow/sale_force_invoiced):

```suggestion
    @api.depends('invoice_blocking_reason_id', 'state', 'order_line.invoice_status')
    def _get_invoice_status(self):
        super()._get_invoice_status()
        for order in self.filtered(
            lambda order: order.invoice_blocking_reason_id and order.state in ("sale", "done")
        ):
            order.invoice_status = "no"
```

**Review by jcoux — CHANGES_REQUESTED** [review]
You need to update translation.

**Review by tarbib — APPROVED** [review]
Functional review :+1: 

https://user-images.githubusercontent.com/2255433/119457550-17ce1000-bd3c-11eb-86f0-80fdb2bd53c4.mp4

**Review by jcoux — APPROVED** [review]
Technically good for me :+1:

**Comment by tarbib** [comment]
@damdam-s Sorry there is an issue actually as I can still invoice while "nothing to invoice" is forced. I should have had the warning message as in standard Odoo which says : "there is nothing to invoice" like : 
image


https://user-images.githubusercontent.com/2255433/119458236-d0944f00-bd3c-11eb-9d71-4ae550820abb.mp4

**Comment by damdam-s** [comment]
@tarbib should be OK now

File: sale_invoice_blocking/models/sale_order.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+            and order.state in ("sale", "done")
+        ):
+            order.invoice_status = "no"
+
+    def _get_invoiceable_lines(self, final=False):
+        """Return the invoiceable lines for order `self`."""
+        if self.invoice_blocking_reason_id:
+            return self.env["sale.order.line"]
+
+        return super(SaleOrder, self)._get_invoiceable_lines(final=final)
```
Comment (jcoux) [inline sale_invoice_blocking/models/sale_order.py]:
```suggestion
        return super()._get_invoiceable_lines(final=final)
```

**Comment by damdam-s** [comment]
@rousseldenis fixed as suggested. can you update your review please?

**Review by rousseldenis — APPROVED** [review]
LGTM

**Review by tarbib — APPROVED** [review]
Functional approval :+1: 
https://user-images.githubusercontent.com/2255433/119478174-e2332200-bd4f-11eb-9ca3-394667f80c63.mp4
@damdam-s it would be better to prevent invoicing block reason creation on the go from the SO though

**Comment by damdam-s** [comment]
rebased

**Review by simahawk — APPROVED** [review]
@damdam-s next time, pls, add a bit of test coverage :pray:

**Pull request merged**